api/src/shop/stripe_event.py

```diff
@@ -92,9 +98,133 @@ def stripe_payment_intent_event(subtype, event):
     else:
         raise IgnoreEvent(f"payment_intent event subtype {subtype} for transaction {transaction.id}")
 
+def stripe_invoice_event(subtype : Subtype, event:any):
+    #FIXME: In the case of uncollectable subtype, we should probably e-mail the invoice to the member
+
+    if (subtype == Subtype.PAID):
+        print(f"Processing paid invoice {event['id']}")
+        # Member has paid something and we can now add things accordingly...
+        invoice = event["data"]['object']
+
+        for line in invoice["lines"]['data']:
+            subscription_id = line['subscription']
+            
+            # Ignore non-subscription lines in invoices
+            if (subscription_id == None):
+                continue
+
+            subscription = stripe.Subscription.retrieve(subscription_id)
+
+            # We ignore any items that doesn't contain the right metadata
+            if ("makerspace_user_id" not in subscription["metadata"] or "subscription_type" not in subscription["metadata"]):
+                print(f'Subscription {subscription_id} does not have metadata indicating an actionable subscription.')
+                continue
+            member_id = int(subscription["metadata"]["makerspace_user_id"])
+            subscription_type = subscription["metadata"]["subscription_type"]
+
+            span_type = None
+            if (subscription_type == SubscriptionTypes.LAB):
+                span_type = Span.LABACCESS
+            elif (subscription_type == SubscriptionTypes.MEMBERSHIP):
+                span_type = Span.MEMBERSHIP
+
+            if (span_type is None):
+                print(f"Invalid subscription type {subscription_type} paid in invoice {invoice['id']}.")
+                continue
+
+            member:Member = db_session.query(Member).get(member_id)
+            
+            if (subscription_type == SubscriptionTypes.LAB):
+                member.stripe_labaccess_subscription_id = subscription_id
+            elif (subscription_type == SubscriptionTypes.MEMBERSHIP):
+                member.stripe_membership_subscription_id = subscription_id
+
+            end_ts = int(line["period"]["end"])
+            start_ts =int(line["period"]["start"])
+
+            # Divide the timespan down to days
+            days = (end_ts-start_ts)/86400
```

Comment:
Does the `add_membership_days` function handle potentially non-integer days?

Reply:
Good point. Probably best to round this to ensure we have integer days.